src/compiler/scala/tools/nsc/transform/async/TransformUtils.scala

```diff
@@ -0,0 +1,308 @@
+
+  lazy val IllegalStateExceptionClass: Symbol = rootMirror.staticClass("java.lang.IllegalStateException")
+  lazy val IllegalStateExceptionClass_NEW_String: Symbol = IllegalStateExceptionClass.info.decl(nme.CONSTRUCTOR).suchThat(
+    x => x.paramss.head.size == 1 && x.firstParam.info.typeSymbol == definitions.StringClass)
+
+  def isAsync(fun: Tree): Boolean = fun.symbol == currentTransformState.ops.Async_async
+  def isAwait(fun: Tree): Boolean = fun.symbol == currentTransformState.ops.Async_await
+
+  def isBooleanShortCircuit(sym: Symbol): Boolean =
+    sym.owner == definitions.BooleanClass && (sym == definitions.Boolean_and || sym == definitions.Boolean_or)
+
+  def isLabel(sym: Symbol): Boolean = sym != null && sym.isLabel
+  def isCaseLabel(sym: Symbol): Boolean = sym != null && sym.isLabel && sym.name.startsWith("case")
+  def isMatchEndLabel(sym: Symbol): Boolean = sym != null && sym.isLabel && sym.name.startsWith("matchEnd")
+
+  def substituteTrees(t: Tree, from: List[Symbol], to: List[Tree]): Tree =
+    (new TreeSubstituter(from, to)).transform(t)
+
+  def statsAndExpr(tree: Tree): (List[Tree], Tree) = tree match {
+    case Block(stats, expr) => (stats, expr)
+    case _                  =>
+      if (tree.tpe <:< definitions.UnitTpe) (Nil, tree)
+      else (List(tree), literalUnit)
+  }
+
+  def listToBlock(trees: List[Tree]): Block = trees match {
+    case trees @ (init :+ last) =>
+      val pos = trees.map(_.pos).reduceLeft(_ union _)
+      Block(init, last).setType(last.tpe).setPos(pos)
+    case Nil =>
+      throw new MatchError(trees)
+  }
+
+  def assignUnitType(t: Tree): t.type = t.setType(definitions.UnitTpe)
+
+  def isUnitType(tp: Type): Boolean = tp.typeSymbol == definitions.UnitClass || tp =:= definitions.BoxedUnitTpe
+
+  def literalUnit: Tree = Literal(Constant(())).setType(definitions.UnitTpe) // a def to avoid sharing trees
+  def literalBoxedUnit: Tree = gen.mkAttributedRef(definitions.BoxedUnit_UNIT)
+
+  def isLiteralUnit(t: Tree): Boolean = t match {
+    case Literal(Constant(())) => true
+    case t if t.symbol == definitions.BoxedUnit_UNIT => true // important to find match labels (which are potential states)
+    case _ => false
+  }
+
+  def adaptToUnit(rhs: List[Tree]): Block = {
+    def filterUnit(ts: List[Tree]): List[Tree] = ts.filterNot(isLiteralUnit(_))
+    rhs match {
+      case (rhs: Block) :: Nil if rhs.tpe <:< definitions.UnitTpe || rhs.tpe <:< definitions.BoxedUnitTpe =>
+        rhs
+      case init :+ last if last.tpe <:< definitions.UnitTpe || last.tpe <:< definitions.BoxedUnitTpe=>
+        Block(filterUnit(init), last).setType(definitions.UnitTpe)
+      case _                                                          =>
+        Block(filterUnit(rhs), literalUnit).setType(definitions.UnitTpe)
+    }
+  }
+
+  private object ThicketAttachment
+  class ThicketTransformer(initLocalTyper: analyzer.Typer) extends TypingTransformer(initLocalTyper) {
```

Comment:
Might be useful outside of async, too. (But we can always move it into a different package later)

Reply:
I've moved it out now, simplified the implenentation a bit, and added a unit test.